Makefile

```diff
@@ -1,25 +1,38 @@
 CXX ?= g++
 CFLAGS = -Wall -Wconversion -O3 -fPIC
 SHVER = 2
-OS = $(shell uname)
+OS := $(shell uname)
 
-all: svm-train svm-predict svm-scale
+all: svm-train svm-predict svm-scale lib
 
-lib: svm.o
-	if [ "$(OS)" = "Darwin" ]; then \
-		SHARED_LIB_FLAG="-dynamiclib -Wl,-install_name,libsvm.so.$(SHVER)"; \
-	else \
-		SHARED_LIB_FLAG="-shared -Wl,-soname,libsvm.so.$(SHVER)"; \
-	fi; \
-	$(CXX) $${SHARED_LIB_FLAG} svm.o -o libsvm.so.$(SHVER)
+ifeq ($(OS),Darwin)
+  DLLEXT := dylib
+  SHARED_LIB_FLAG = -dynamiclib
+else
+  DLLEXT := so
+  SHARED_LIB_FLAG = -shared
+endif
+
+libsvm.$(DLLEXT): svm.o
+	$(CXX) $(SHARED_LIB_FLAG) $^ -o $@
+
+libsvm.$(DLLEXT).$(SHVER): libsvm.$(DLLEXT)
+	ln $< $@
+
+libsvm.a: svm.o
+	ar rc $@ $^ 
+	ranlib $@ 
+
+.PHONY: lib
+lib: libsvm.$(DLLEXT).$(SHVER) libsvm.a
 
 svm-predict: svm-predict.c svm.o
-	$(CXX) $(CFLAGS) svm-predict.c svm.o -o svm-predict -lm
+	$(CXX) $(CFLAGS) $^ -o $@ -lm
 svm-train: svm-train.c svm.o
-	$(CXX) $(CFLAGS) svm-train.c svm.o -o svm-train -lm
+	$(CXX) $(CFLAGS) $^ -o $@ -lm
 svm-scale: svm-scale.c
-	$(CXX) $(CFLAGS) svm-scale.c -o svm-scale
+	$(CXX) $(CFLAGS) $^ -o svm-scale
 svm.o: svm.cpp svm.h
-	$(CXX) $(CFLAGS) -c svm.cpp
+	$(CXX) $(CFLAGS) -c $<
```

Comment:
out of curiosity: Couldn't we use [implicit make rules](https://ftp.gnu.org/old-gnu/Manuals/make-3.79.1/html_chapter/make_10.html#SEC95) for the 4 targets above?